server/server/src/file_service.rs

```diff
@@ -1,213 +1,242 @@
     let (request, server_state) = (&context.request, context.server_state);
-    let mut transaction = match server_state.index_db_client.begin().await {
-        Ok(t) => t,
-        Err(e) => {
-            return Err(InternalError(format!("Cannot begin transaction: {:?}", e)));
-        }
-    };
+    let mut con = server_state.index_db_pool.get().await?;
 
-    let result = file_index_repo::change_document_version_and_size(
-        &mut transaction,
-        request.id,
-        request.new_content.value.len() as u64,
-        request.old_metadata_version,
-    )
-    .await;
+    let watched_keys = &[file(request.id), size(request.id)];
+    let mut new_version = 0;
+    let mut old_content_version = 0;
 
-    let (old_content_version, new_version) = result.map_err(|e| match e {
-        file_index_repo::ChangeDocumentVersionAndSizeError::DoesNotExist => {
-            ClientError(ChangeDocumentContentError::DocumentNotFound)
-        }
-        file_index_repo::ChangeDocumentVersionAndSizeError::IncorrectOldVersion => {
-            ClientError(ChangeDocumentContentError::EditConflict)
+    let tx = tx!(&mut con, pipe, watched_keys, {
+        let new_size = FileUsage {
+            file_id: request.id,
+            size_bytes: request.new_content.value.len() as u64,
+        };
+        new_version = get_time().0 as u64;
+        let mut meta: EncryptedFileMetadata =
+            con.maybe_json_get(file(request.id))
+                .await?
+                .ok_or(Abort(ClientError(
+                    ChangeDocumentContentError::DocumentNotFound,
+                )))?;
+
+        if meta.deleted {
+            return Err(Abort(ClientError(
+                ChangeDocumentContentError::DocumentDeleted,
+            )));
         }
-        file_index_repo::ChangeDocumentVersionAndSizeError::Deleted => {
-            ClientError(ChangeDocumentContentError::DocumentDeleted)
+
+        if meta.owner.is_empty() {
+            return Err(Abort(ClientError(
+                ChangeDocumentContentError::NotPermissioned,
+            )));
         }
-        file_index_repo::ChangeDocumentVersionAndSizeError::Postgres(_)
-        | file_index_repo::ChangeDocumentVersionAndSizeError::Deserialize(_) => {
-            InternalError(format!(
-                "Cannot change document content version in Postgres: {:?}",
-                e
-            ))
+
+        if request.old_metadata_version != meta.metadata_version {
+            return Err(Abort(ClientError(ChangeDocumentContentError::EditConflict)));
         }
-    })?;
 
-    let create_result = file_content_client::create(
+        old_content_version = meta.content_version;
+
+        meta.content_version = new_version;
+        meta.metadata_version = new_version;
+
+        pipe.json_set(size(request.id), new_size)?
+            .json_set(file(request.id), meta)
+    });
+    return_if_error!(tx);
+
+    file_content_client::create(
```

Comment:
the metadata update should happen in-between creating the new version of the document and deleting the old version (see https://github.com/lockbook/lockbook/pull/938/files#diff-4a07b1d902803e753277b5b5c01e349a3960b5023e91d000e4a448187bdbccccR168)

Reply:
fixed: 035d9ef

but 

we're gonna wanna do something better for sure, we eat 2 s3 interactions for every doc change, ~50-150ms each.

We can reduce it down to 1 if we process the cleanup asynchronously, throw it on a redis queue and have a background worker churn on them.

We can reduce it down to 0 if we write to redis exclusively, mark the keys as non-volatile until a worker persists them to s3, and then tell redis, if memory is running low you can lru-evict this key.